Coverage test sys.settrace & improve coverage

jepler opened the pull request:
### Summary

I noticed that `py/profile.c` was not covered at all according to codecov, due to the fact it was never enabled in the unix port.

I enabled it, and then added a new test to cover some previously un-covered lines.

### Testing

I ran the tests locally and reviewed the coverage of profile.c.

Note: I suspect one or more Windows x64 builds will fail **due to improved coverage in this PR**, exposing a latent problem.

I wrote a [rambling post in the discussions area about passing qstrs to mp_printf](https://github.com/orgs/micropython/discussions/17537). The TL;DR explanation is, there appears to be undefined behavior that occurs when a "small qstr" (`uint16_t`) argument such as `type->name` is printed via the `%q` format specifier, which retrieves a `ssize_t` value. The specific behavior I saw with x64 msvc on godbolt looked like it would trigger on one specific format string in the core, the repr of frame objects. That's what spurred me to ensure that this code was covered.

If the failure does occur, I'll try to write it up properly as an issue.

jepler committed:
coverage: Enable sys.settrace.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
tests: Improve test coverage of py/profile.c.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
sys_settrace_features: Add expected output.

This side-steps the problem where the output does not match when the
host python version is 3.12.x (as it is on unix ci today).

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented:
While there are some other errors going on that also leave me scratching my head, the "expected error" happens on Windows CI, specifically "build-vs (x64, Debug, dev, 2017)".

```
Run python run-tests.py --print-failures
--- D:/a/micropython/micropython/tests/results\misc_sys_settrace_cov.py.exp	2025-06-20 18:41:30.413152600 +0000
+++ D:/a/micropython/micropython/tests/results\misc_sys_settrace_cov.py.out	2025-06-20 18:41:30.413152600 +0000
@@ -1,2 +1,2 @@
-FRAME <frame at 0x\[0-9a-f\]\+, file 'misc/sys_settrace_repr\.py', line \\d\+, code f>
-LASTI \\d\+
+FRAME <frame at 0xe06df4a0, file 'D:/a/micropython/micropython/tests/misc/sys_settrace_cov.py', line 20, code Assertion failed: *q < pool->len, file d:\a\micropython\micropython\py\qstr.c, line 198
+CRASH
\ No newline at end of file

FAILURE D:/a/micropython/micropython/tests/results\misc_sys_settrace_cov.py
```

jepler commented:
I can't come up with any good theory as to what's making those unix builds fail. It's also a surprise that it's only ONE of the msvc x64 builds.

Josverl commented:
While working on extending sys.settrace() with additional capabilities I also noticed that the debug builds are not tested across ports. I first noticed Errors first popping up when building an ESP32 locally.

Is there a way to determine coverage locally?

jepler committed:
runtime: Initialize fields in mp_thread_init_state.

If the fields added for MICROPY_PY_SYS_SETTRACE are not
initialized properly, their value in a thread is
indeterminate. In particular, if the callback is not NULL,
it will be invoked as a function.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
coverage: Initialize more code_state fields.

When MICROPY_PY_SYS_SETTRACE was enabled, a crash was seen in
the qemu_mips build. It seems likely that this was due to these
added fields not being initialized.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented:
Yes, you can go through the steps of the unix coverage build locally. In `ports/unix`, `make VARIANT=coverage test_gcov`. The build output will finish with a summary of coverage, and a large number of ".gcov" files will be written. 

gcov files show coverage on a line by line basis, with "#####" indicating no coverage of the line, "-" indicating no coverage possible, and a number indicating the number of times the line was covered. The format is described in `man gcov`.

jepler committed:
profile: Fix printing frame objects.

The argument corresponding to a `%q` specifier must be of type
`qstr`, not a narrower type like `int16_t`. Not ensuring this
caused an assertion error on one Windows x64 build.

The argument corresponding to a `%d` specifier must be of type
`int`, not a potentially-wider type like `mp_uint_t`. Not ensuring
this prevented the function name from being printed on the
unix nanbox build.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented:
The nanbox build were failing like this:
```
-FRAME <frame at 0x\[0-9a-f\]\+, file '\.\*/sys_settrace_cov.py', line \\d\+, code f>
+FRAME <frame at 0xf4d622e0, file 'misc/sys_settrace_cov.py', line 16, code >
```
The reason is that a `mp_uint_t` is passed in to a `%d` format specifier. But in nanbox, `sizeof(mp_uint_t) > sizeof(int)`, so again there's the UB problem that arises when the promoted argument type does not match the va_arg type.

jepler commented:
This PR is now read only because it decreases the overall coverage percentage. In reality, it adds coverage to 21751 previously not covered lines.

jepler commented:
I've updated the PR so that settrace is only enabled in the coverage build, not the standard build or other variants.

we should consider whether this makes the standard+settrace CI build redundant. I can remove that if desired. the standard+stackless+settrace CI job is probably still useful.

Josverl commented:
wrt the work on `micropython-lib/python-ecosys/debugpy` I have been contemplating a few changes as well 
- moving the `misc/sys_settrace*.*` tests to their own folder: `tests/sys_settrace`
- there I have been updating and adding tests to cover the new functionality @andrewleech and I have been working on 

That makes it simpler to run that subset during development.

Also if you have guidance how to test the combination of a sys,settrace firmware together with `micropython-lib/python-ecosys/debugpy` that would be appreciated. Creating a separate 'settrace/debug' variant seems the simplest approach to me.

jepler committed:
ci: Remove the "settrace" build.

This becomes redundant when the main coverage build includes
settrace.

Signed-off-by: Jeff Epler <jepler@gmail.com>

dpgeorge commented:
> * moving the `misc/sys_settrace*.*` tests to their own folder: `tests/sys_settrace`

That sounds reasonable, although not in scope for this PR.

> * Also if you have guidance how to test the combination of a sys,settrace firmware together with `micropython-lib/python-ecosys/debugpy` that would be appreciated. Creating a separate 'settrace/debug' variant seems the simplest approach to me.

This PR will enable settrace on the unix coverage build, so once this PR is merged you could use that.

jepler committed:
tests: Increase test timeout in coverage build.

The additional overhead of the settrace profiler means that the
`aes_stress.py` test was running too slowly on GitHub CI.
Double the timeout to 60 seconds.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler commented:
I added an environment variable override and set it from ci.sh when running coverage tests. I tested that locally, by making sure that a timeout of 2 produced a handful of failures.

jepler committed:
ci: Use Python 3.11 for coverage testing.

This removes the need for the `sys_settrace_features.py.exp` file.

This means that people testing locally will also need to install
Python 3.11 in some way, such as with pyenv or uv, and use it
during `make VARIANT=coverage test`, or they will get failures.

When using python from GitHub actions/setup-python, pip3 can't
be wrapped by sudo, because this invokes the operating system python
instead.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
objcode: Fully parenthesize macro expansion.

as requested in code review.

Signed-off-by: Jeff Epler <jepler@gmail.com>

jepler committed:
runtime: Initialize bool field idiomatically.

Signed-off-by: Jeff Epler <jepler@gmail.com>

dpgeorge commented:
Let me know when this is ready.

dpgeorge reviewed (approved):
Looking good now!

dpgeorge commented:
Squashed the fix-up commits and merged in fcfed6a0ea763ff14ea2bf819179fdf7eecda889 through a9801f9960ea2b8e94d5626840c97bc45286e1a3

jepler commented:
Thank you!
